### PR TITLE
[Security][WIP] Add isSafeShell check before passing extra params to mail()

### PR DESCRIPTION
**Pull request by barryvdh**
Potential fix for #844

@Zenexer Are you okay with using the (slightly reformatted) code from https://github.com/PHPMailer/PHPMailer/pull/929 here? Do you agree this fixes the issue?

### CODE REVIEW

**Comment by Zenexer** [comment]
@barryvdh Checking now.

File: lib/classes/Swift/Transport/MailTransport.php [hunk 1]
```diff
@@ -239,6 +239,35 @@ private function _getReversePath(Swift_Mime_Message $message)
     }
 
     /**
+     * Fix CVE-2016-10074 by disallowing potentially unsafe shell characters.
+     *
+     * Note that escapeshellarg and escapeshellcmd are inadequate for our purposes, especially on Windows.
+     * @param string $string The string to be validated
+     * @see https://github.com/swiftmailer/swiftmailer/issues/844
+     * @return boolean
+     */
```
Comment (fabpot) [inline lib/classes/Swift/Transport/MailTransport.php]: We need to ask permission to reuse this piece of code as PHPMailer is released under the GPL license, whereas we are using MIT for Swiftmailer.

File: lib/classes/Swift/Transport/MailTransport.php [hunk 2]
```diff
@@ -249,7 +278,9 @@ private function _getReversePath(Swift_Mime_Message $message)
     private function _formatExtraParams($extraParams, $reversePath)
     {
         if (false !== strpos($extraParams, '-f%s')) {
-            // no need to escape $reversePath) as mail() already does it
+            if (false === $this->_isShellSafe($reversePath)) {
```
Comment (Zenexer) [inline lib/classes/Swift/Transport/MailTransport.php]: Unlike on line 200, there's no need to explicitly check for `false` here.  `if (!$this->_isShellSafe($reversePath)` should be sufficient.

File: lib/classes/Swift/Transport/MailTransport.php (hunk 1 above)
Comment (Zenexer) [inline lib/classes/Swift/Transport/MailTransport.php]: I'm going to be releasing my own version under public domain shortly.

File: lib/classes/Swift/Transport/MailTransport.php [hunk 3]
```diff
@@ -249,7 +278,9 @@ private function _getReversePath(Swift_Mime_Message $message)
...
+                $reversePath = null;
+            }
             $extraParams = empty($reversePath) ? str_replace('-f%s', '', $extraParams) : sprintf($extraParams, $reversePath);
```
Comment (fabpot) [inline lib/classes/Swift/Transport/MailTransport.php]: I would do:

```php
            if (empty($reversePath) || false === $this->_isShellSafe($reversePath)) {
                $extraParams = str_replace('-f%s', '', $extraParams);
            } else {
                $extraParams = sprintf($extraParams, $reversePath);
            }
```

File: lib/classes/Swift/Transport/MailTransport.php (hunk 1 above)
Comment (barryvdh) [inline lib/classes/Swift/Transport/MailTransport.php]: I asked for permission by @Zenexer who is the original author. Would that not suffice?

File: lib/classes/Swift/Transport/MailTransport.php (hunk 2 above)
Comment (fabpot) [inline lib/classes/Swift/Transport/MailTransport.php]: Here, it's more a coding standard than a strict requirement.

File: lib/classes/Swift/Transport/MailTransport.php (hunk 3 above)
Comment (barryvdh) [inline lib/classes/Swift/Transport/MailTransport.php]: I changed this as suggested

File: lib/classes/Swift/Transport/MailTransport.php (hunk 2 above)
Comment (barryvdh) [inline lib/classes/Swift/Transport/MailTransport.php]: I changed it, but do you prefer it back as `false !== ` @fabpot ?

File: lib/classes/Swift/Transport/MailTransport.php [hunk 4]
```diff
@@ -239,6 +239,38 @@ private function _getReversePath(Swift_Mime_Message $message)
     }
 
     /**
+     * Fix CVE-2016-10074 by disallowing potentially unsafe shell characters.
+     *
+     * Note that escapeshellarg and escapeshellcmd are inadequate for our purposes, especially on Windows.
+     *
+     * @param string $string The string to be validated
+     *
+     * @see https://github.com/swiftmailer/swiftmailer/issues/844
```
Comment (fabpot) [inline lib/classes/Swift/Transport/MailTransport.php]: This one can be removed.

File: tests/unit/Swift/Transport/MailTransportTest.php [hunk 5]
```diff
@@ -391,7 +418,7 @@ public function testMessageHeadersOnlyHavePHPEolsDuringSending()
     }
 
     /**
-     * @expectedException Swift_TransportException
+     * @expectedException \Swift_TransportException
```
Comment (fabpot) [inline tests/unit/Swift/Transport/MailTransportTest.php]: should be reverted, we don't have namespace in Switfmailer

File: lib/classes/Swift/Transport/MailTransport.php [hunk 6]
```diff
@@ -249,8 +281,11 @@ private function _getReversePath(Swift_Mime_Message $message)
     private function _formatExtraParams($extraParams, $reversePath)
     {
         if (false !== strpos($extraParams, '-f%s')) {
-            // no need to escape $reversePath) as mail() already does it
-            $extraParams = empty($reversePath) ? str_replace('-f%s', '', $extraParams) : sprintf($extraParams, $reversePath);
+            if (empty($reversePath) || !$this->_isShellSafe($reversePath)) {
```
Comment (fabpot) [inline lib/classes/Swift/Transport/MailTransport.php]: Indeed, I would prefer to keep `false ===`

File: tests/unit/Swift/Transport/MailTransportTest.php (hunk 5 above)
Comment (barryvdh) [inline tests/unit/Swift/Transport/MailTransportTest.php]: Sorry, your fabbot.io told me to, I can revert but woould probably trigger a false check on fabbot, see the patch here: http://fabbot.io/patch/swiftmailer/swiftmailer/846/aa17624da761b7a12ffa46590be1f06ed7ae9c4a/cs.diff

Comment (fabpot) [inline tests/unit/Swift/Transport/MailTransportTest.php]: I know, that's an issue with PHP CS Fixer :(

Comment (barryvdh) [inline tests/unit/Swift/Transport/MailTransportTest.php]: Ok, changed

**Comment by barryvdh** [comment]
@Zenexer Can you explicitly give permission to use you code in Swiftmailer under the MIT license?

**Comment by fabpot** [comment]
Thanks @barryvdh. If @Zenexer is ok for us to use his snippet of code, I will go ahead and release a new version of Swiftmailer.

Actually, I will do one more thing: deprecate the Mail transport and remove it in 6.0.

**Comment by fabpot** [comment]
905b0bc deprecates the mail transport. The reasoning being that as `mail()` has many drawbacks besides security (as mentioned in the phpdocs) and because this is just a wrapper on top of `sendmail`, there is no need to keep it.

**Comment by fabpot** [comment]
@barryvdh Can you add a note in the `CHANGES` file?

**Comment by fabpot** [comment]
Thank you @barryvdh.

**Comment by barryvdh** [comment]
Whoops, was a minute to late with my changes

**Comment by Zenexer** [comment]
[This variation](https://gist.github.com/Zenexer/40d02da5e07f151adeaeeaa11af9ab36#sanitizing-input-for-safe-shell-parsing) is now in the public domain and has a few additional sanity checks to make it more generic.

It's worth noting that using sendmail directly through popen is also insecure.  The only way to securely start a process while passing user input as an argument is with pcntl_fork + pcntl_exec because it uses execve instead of the shell.

**Comment by Zenexer** [comment]
Yup, this is what happens when I code at 5 AM.  Nice catch. :)

**Comment by Zenexer** [comment]
Fixed.  It was meant to be `if (!$length)`.

You can use it without attribution under whatever license you like.  Just keep in mind that people can copy that portion of code without being bound by some restrictions of MIT; for example, I believe they can copy any minor modifications you make without providing attribution.

**Comment by barryvdh** [comment]
> 905b0bc deprecates the mail transport. The reasoning being that as mail() has many drawbacks besides security (as mentioned in the phpdocs) and because this is just a wrapper on top of sendmail, there is no need to keep it.

@fabpot This is not the case for Windows, right? There it's a wrapper for SMTP or an alternative sendmail binary (if set in the php.ini), so not 100% drop in (eg. same could would work with MailTransport on windows + linux without application config (is php is configured to handle mail properly), but sendmail would only work on Linux right?

**Comment by fabpot** [comment]
If it's a wrapper for SMTP, then the same argument applies, just use the SMTP server. The point being that whatever the configuration, mail() relies on SMTP or sendmail and we already have support for those.

**Pull request closed without merging**